### PR TITLE
auth LUA: add ifurlextup function

### PR DESCRIPTION
**Pull request by Habbie**
### Short description
Still needs tests.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ]  checked that this code was merged to master

### CODE REVIEW

**Comment by neilcook** [comment]
I would like for the "grouping of IPs" concept that is in ifurlup to be added to the PR if possible. Then ifurlhcup would be functionally equivalent to ifurlup, just with different syntax and a different way to get the healthcheck information.

**Comment by Habbie** [comment]
Pending tests, example usage: `z.example.com   3005    IN      LUA     A "ifurlextup({{'192.168.0.1', 'https://7bits.nl/404'},{'192.168.0.2', 'https://example.com/'}})"`

**Comment by Habbie** [comment]
Equivalent with grouping: `"ifurlextup({{['192.168.0.1']='https://7bits.nl/404',['192.168.0.2']='https://example.com/404'}})"`

**Comment by Habbie** [comment]
two groups: `"ifurlextup({{['192.168.0.1']='https://7bits.nl/404',['192.168.0.2']='https://example.com/404'}, {['192.168.0.3']='https://example.net/'}})"`

**Comment by neilcook** [comment]
> two groups: `"ifurlextup({{['192.168.0.1']='https://7bits.nl/404',['192.168.0.2']='https://example.com/404'}, {['192.168.0.3']='https://example.net/'}})"`

The docs need updating to reflect this new syntax of course

**Comment by neilcook** [comment]
Docs section LGTM

**Review by omoerbeek — APPROVED** [review]
Looks good. With two general comments:
- no spaces after `conditional_keyword` and `(`
- This is more of a feature request than a review comment.... I would love to be able to return an IP from a _specific_ set if all tests fail. AFAKS, that is not possible.

**Review by rgacogne — APPROVED** [review]
Looks good to me. I made one code readability suggestion but YMMV :-)

**Comment by Habbie** [comment]
> * I would love to be able to return an IP from a _specific_ set if all tests fail. AFAKS, that is not possible.

Indeed, this is not possible. I have in fact been pondering the right design for this over the last few days, I'll likely PR something for that, separately, soon.

File: docs/lua-records/functions.rst [hunk 1]
```diff
@@ -90,7 +90,31 @@ Record creation functions
...
+
+    ifurlextup({{['192.168.0.1']='https://example.com/',['192.168.0.2']='https://example.com/404'}})
+
+  Example with two groups:
+
+  .. code-block:: lua
+
+    ifurlextup({{['192.168.0.1']='https://example.net/404',['192.168.0.2']='https://example.com/404'}, {['192.168.0.3']='https://example.net/'}})"
+
+  The health checker will look up the first two URLs (using normal DNS resolution to find them - whenever possible, use URLs with IPs in them).
```
[the remark this answers is not in the document]
Comment (Habbie) [reply docs/lua-records/functions.rst:115]: then an up member from the first group is returned, if one exists

(after external conversion, it's clear that the wording is not great, for `ifurlup` as well, we'll fix that separately)

**Comment by Habbie** [comment]
> no spaces after `conditional_keyword` and `(`

Looks like Rémi fixed that for us :)

**Pull request merged**